### PR TITLE
Basic resilient communication

### PR DESCRIPTION
**Pull request by Anghul**
Addition of a functionnal communication between servers.
A few things may not be clear just by seeing them :
- addition of the socket in the publish method for the channels : it allows a server to catchup on a channel only when the channel was created by a message received by a server. It is needed as an election channel is created by a publish message on a lao
- addition of the public key of the organizer in the lao channels : needed as only the organizer is allowed to do certain operations, and only the organizer server would know who the organizer without it

### CODE REVIEW

File: be1-go/channel/lao/mod.go [hunk 1]
```diff
@@ -418,17 +437,17 @@ func (c *Channel) processRollCallObject(action string, msg message.Message) erro
...
 	// Check if the sender of the roll call message is the organizer
 	senderPoint := crypto.Suite.Point()
 	err = senderPoint.UnmarshalBinary(senderBuf)
 	if err != nil {
-		return answer.NewErrorf(-4, "failed to unmarshal public key of the sender: %v", err)
+		return answer.NewErrorf(-4, keyUnmarshalError, err)
 	}
 
-	if !c.hub.GetPubkey().Equal(senderPoint) {
+	if !c.organizerKey.Equal(senderPoint) {
```
Comment (celinecamacho) [inline be1-go/channel/lao/mod.go:450]: So now we have the organizer key in the channel because the `hub.GetPubkey()` might give the public key of a witness?

File: be1-go/cli/organizer/organizer.go [hunk 2]
```diff
@@ -20,12 +24,13 @@ func Serve(cliCtx *cli.Context) error {
...
 
 	// get command line args which specify public key, port to use for clients
-	// and witnesses
+	// and witnesses, witness' address
 	clientPort := cliCtx.Int("client-port")
 	witnessPort := cliCtx.Int("witness-port")
 	if clientPort == witnessPort {
 		return xerrors.Errorf("client and witness ports must be different")
 	}
+	witness := cliCtx.StringSlice("other-witness")
```
Comment (celinecamacho) [inline be1-go/cli/organizer/organizer.go:33]: Maybe verify that the witness is not empty, just as for the pk in line 36

File: be1-go/cli/utility/mod.go [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+	"popstellar/network/socket"
+	"sync"
+
+	"github.com/gorilla/websocket"
+	"golang.org/x/xerrors"
+)
+
+// ConnectToSocket establishes a connection to another server's witness
+// endpoint.
+func ConnectToSocket(otherHubType hub.HubType, address string, h hub.Hub, wg *sync.WaitGroup, done chan struct{}) error {
```
Comment (celinecamacho) [inline be1-go/cli/utility/mod.go:18]: Why did you delete this function from the witness package and added it to a utility one?

File: be1-go/hub/hub/message_handling.go [hunk 4]
```diff
@@ -0,0 +1,287 @@
+package hub
```
Comment (celinecamacho) [inline be1-go/hub/hub/message_handling.go:1]: It seems weird to have to packages named hub, one inside the other: `hub/hub`. Maybe you can call the second package `general`

**Review by celinecamacho — COMMENTED** [review]
Great job. I have a few comments, I still need to review the other half of the files :)

**Review by rjaccard — CHANGES_REQUESTED** [review]
Whouaou! Big PR! Nice work.

File: be1-go/channel/lao/mod.go (hunk 1 above)
Comment (Anghul) [reply be1-go/channel/lao/mod.go:450]: Exactly, as hub.GetPubkey() gives the public key of the front-end linked to the current back-end, that could be a witness

File: be1-go/cli/organizer/organizer.go (hunk 2 above)
Comment (Anghul) [reply be1-go/cli/organizer/organizer.go:33]: No, I don't think we want to force the use of multiple back ends for now, and even if we wanted to force it, the first server to be launched need to have this field empty to avoid an error to be thrown

File: be1-go/cli/utility/mod.go (hunk 3 above)
Comment (Anghul) [reply be1-go/cli/utility/mod.go:18]: Exactly

File: be1-go/hub/hub/mod.go [hunk 5]
```diff
@@ -151,68 +176,71 @@ func (h *Hub) Receiver() chan<- socket.IncomingMessage {
...
+	return nil
+}
+
+// OnSocketClose implements hub.Hub
+func (h *Hub) OnSocketClose() chan<- string {
+	return h.closedSockets
+}
+
+func (h *Hub) getChan(channelPath string) (channel.Channel, error) {
+	if channelPath[:6] != rootPrefix {
```
[the remark this answers is not in the document]
Comment (Anghul) [reply be1-go/hub/hub/mod.go:231]: It still need to be equal to /root/, even if the const is changed

File: be1-go/network/socket/socket.go [hunk 6]
```diff
@@ -89,7 +89,7 @@ func (s *baseSocket) ReadPump() {
 					Str("socket", s.conn.RemoteAddr().String()).
 					Msg("connection dropped unexpectedly")
 			} else {
-				s.log.Info().Msg("closing the read pump")
+				s.log.Err(err).Msg("closing the read pump")
```
[the remark this answers is not in the document]
Comment (Anghul) [reply be1-go/network/socket/socket.go:92]: It is just an artifact from when I was testing

File: be1-go/hub/hub/mod.go [hunk 7]
```diff
@@ -1,12 +1,12 @@
-package organizer
+package hub
```
[the remark this answers is not in the document]
Comment (Anghul) [reply be1-go/hub/hub/mod.go:1]: We still need the interface for the fakeHubs used during the tests

File: be1-go/hub/hub/mod.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (Anghul) [reply be1-go/hub/hub/mod.go]: It still need to be equal to /root/, even if the constant is changed

File: be1-go/hub/hub/mod.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (Anghul) [reply be1-go/hub/hub/mod.go]: We still need the interface to create fakeHub for the tests

File: be1-go/cli/utility/mod.go (hunk 3 above)
Comment (Anghul) [reply be1-go/cli/utility/mod.go:18]: Exactly

File: be1-go/channel/lao/mod.go (hunk 1 above)
Comment (Anghul) [reply be1-go/channel/lao/mod.go:451]: Yes, exactly

**Review by rjaccard — DISMISSED** [review]
:)

**Review by celinecamacho — APPROVED** [review]

File: be1-go/hub/standard_hub/mod.go [hunk 8]
```diff
@@ -151,68 +181,71 @@ func (h *Hub) Receiver() chan<- socket.IncomingMessage {
...
-// handleRootChannelMesssage handles an incoming message on the root channel.
-func (h *Hub) handleRootChannelMesssage(socket socket.Socket, publish method.Publish) error {
-	jsonData, err := base64.URLEncoding.DecodeString(publish.Params.Message.Data)
-	if err != nil {
-		err := xerrors.Errorf("failed to decode message data: %v", err)
-		socket.SendError(&publish.ID, err)
-		return err
-	}
+// CatchupToServer sends a catchup query to another server
+func (h *Hub) CatchupToServer(socket socket.Socket, channel string) error {
```
[the remark this answers is not in the document]
Comment (nkcr) [reply be1-go/hub/standard_hub/mod.go:193]: I would delegate the responsibility to create the catchup message to the `queries` struct:

```go
func (q *queries) GetNextCatchupMsg(channel string) method.Catchup {
	q.Lock()
	defer q.Unlock()

	catchupID := q.nextID

	baseValue := false
	q.state[catchupID] = &baseValue

	rpcMessage := method.Catchup{
		Base: query.Base{
			JSONRPCBase: jsonrpc.JSONRPCBase{
				JSONRPC: "2.0",
			},
			Method: "catchup",
		},
		ID: catchupID,
		Params: struct {
			Channel string "json:\"channel\""
		}{
			channel,
		},
	}

	q.queries[catchupID] = rpcMessage
	q.nextID ++

	return rpcMessage
}
```

File: be1-go/hub/standard_hub/mod.go [hunk 9]
```diff
@@ -63,10 +67,31 @@ type Hub struct {
...
+	// hubInbox is used to remember which messages were broadcast by the
+	// server to avoid broadcast loops
+	hubInbox inbox.Inbox
+
+	// rootInbox and queries are used to help servers catchup to each other
+	rootInbox inbox.Inbox
+	queries   queries
+}
+
+type queries struct {
```
Comment (nkcr) [inline be1-go/hub/standard_hub/mod.go:82]: please comment why the type is needed and how it is used

File: be1-go/hub/standard_hub/mod.go [hunk 10]
```diff
@@ -63,10 +67,31 @@ type Hub struct {
...
+	mu sync.Mutex
+	// state store the ID of the server queries and their state. False for a
+	// query not yet answered, else true.
+	state   map[int]*bool
+	queries map[int]method.Catchup
+
+	// nextID store the ID of the next query
```
Comment (nkcr) [inline be1-go/hub/standard_hub/mod.go:89]: please explain why it is needed and how it is used

File: be1-go/hub/standard_hub/mod.go [hunk 11]
```diff
@@ -63,10 +67,31 @@ type Hub struct {
...
+	nextID int
 }
 
-// NewHub returns a Organizer Hub.
-func NewHub(public kyber.Point, log zerolog.Logger, laoFac channel.LaoFactory) (*Hub, error) {
+// NewHub returns a new Witness Hub.
+func NewHub(public kyber.Point, log zerolog.Logger, laoFac channel.LaoFactory, hubType hub.HubType) (*Hub, error) {
```
Comment (nkcr) [inline be1-go/hub/standard_hub/mod.go:94]:
```suggestion
func NewHub(public kyber.Point, log zerolog.Logger, laoFac channel.LaoFactory, 
	hubType hub.HubType) (*Hub, error) {
```

File: be1-go/hub/standard_hub/mod.go [hunk 12]
```diff
@@ -85,6 +111,10 @@ func NewHub(public kyber.Point, log zerolog.Logger, laoFac channel.LaoFactory) (
 		workers:         semaphore.NewWeighted(numWorkers),
 		log:             log,
 		laoFac:          laoFac,
+		serverSockets:   channel.NewSockets(),
+		hubInbox:        *inbox.NewInbox(rootChannel),
+		rootInbox:       *inbox.NewInbox(rootChannel),
+		queries:         queries{state: make(map[int]*bool), queries: make(map[int]method.Catchup), nextID: 0},
```
Comment (nkcr) [inline be1-go/hub/standard_hub/mod.go:117]: declare and use a `newQueries` constructor instead. This constructor must be right above the struct declaration.

File: be1-go/hub/standard_hub/mod.go [hunk 13]
```diff
@@ -151,68 +181,71 @@ func (h *Hub) Receiver() chan<- socket.IncomingMessage {
...
+	return nil
+}
+
+// OnSocketClose implements hub.Hub
+func (h *Hub) OnSocketClose() chan<- string {
+	return h.closedSockets
+}
+
+func (h *Hub) getChan(channelPath string) (channel.Channel, error) {
+	if channelPath[:6] != rootPrefix {
```
Comment (nkcr) [inline be1-go/hub/standard_hub/mod.go:236]: Use `strings.HasPrefix()`. This check can panic.

File: be1-go/hub/standard_hub/mod.go [hunk 14]
```diff
@@ -151,68 +181,71 @@ func (h *Hub) Receiver() chan<- socket.IncomingMessage {
...
+		return nil, xerrors.Errorf("channel id must begin with \"/root/\", got: %q", channelPath[:6])
```
Comment (nkcr) [inline be1-go/hub/standard_hub/mod.go:237]:
```suggestion
		return nil, xerrors.Errorf("channel not prefixed with '%s': %q", rootPrefix, channelPath)
```

File: be1-go/hub/standard_hub/mod.go [hunk 15]
```diff
@@ -418,29 +406,42 @@ func (h *Hub) handleIncomingMessage(incomingMessage *socket.IncomingMessage) err
...
+		h.serverSockets.SendToAll(byteMessage)
+		h.hubInbox.StoreMessage(message)
+		return false
+	}
+	return true
+}
+
 // createLao creates a new LAO using the data in the publish parameter.
-func (h *Hub) createLao(publish method.Publish, laoCreate messagedata.LaoCreate) error {
+func (h *Hub) createLao(publish method.Publish, laoCreate messagedata.LaoCreate, socket socket.Socket) error {
```
Comment (nkcr) [inline be1-go/hub/standard_hub/mod.go:432]:
```suggestion
func (h *Hub) createLao(publish method.Publish, laoCreate messagedata.LaoCreate,
	socket socket.Socket) error {
```

File: be1-go/hub/standard_hub/mod.go [hunk 16]
```diff
@@ -460,16 +461,23 @@ func (h *Hub) GetSchemaValidator() validation.SchemaValidator {
 }
 
 // RegisterNewChannel implements channel.HubFunctionalities
-func (h *Hub) RegisterNewChannel(channeID string, channel channel.Channel) {
+func (h *Hub) RegisterNewChannel(channelID string, channel channel.Channel, sock socket.Socket) {
```
Comment (nkcr) [inline be1-go/hub/standard_hub/mod.go:464]: Should be renamed to `RegisterAndCatchupChan`, or `NotifyNewChannel` (better).

File: be1-go/network/server.go [hunk 17]
```diff
@@ -123,6 +123,9 @@ func (s *Server) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 
 		go witness.ReadPump()
 		go witness.WritePump()
+
+		err = s.h.AddServerSocket(witness)
+		s.log.Err(err).Msg("error while trying to catchup server")
```
Comment (nkcr) [inline be1-go/network/server.go:128]: Add:

```go
		http.Error(w, "failed to add socket: " + err.Error(), http.StatusInternalServerError)
		return
```

File: be1-go/hub/standard_hub/mod.go [hunk 18]
```diff
@@ -151,68 +181,71 @@ func (h *Hub) Receiver() chan<- socket.IncomingMessage {
...
 
-// OnSocketClose implements hub.Hub
-func (h *Hub) OnSocketClose() chan<- string {
-	return h.closedSockets
+// AddServerSocket adds a socket to the sockets known by the hub and query a
+// catchup from the new server
+func (h *Hub) AddServerSocket(socket socket.Socket) error {
+	h.serverSockets.Upsert(socket)
+
+	return h.CatchupToServer(socket, rootChannel)
```
Comment (nkcr) [inline be1-go/hub/standard_hub/mod.go:189]: Returning a function call makes the code harder to follow. Please do it in two steps.

File: be1-go/hub/standard_hub/mod.go [hunk 19]
```diff
@@ -151,68 +181,71 @@ func (h *Hub) Receiver() chan<- socket.IncomingMessage {
 	return h.messageChan
 }
 
-// OnSocketClose implements hub.Hub
-func (h *Hub) OnSocketClose() chan<- string {
-	return h.closedSockets
+// AddServerSocket adds a socket to the sockets known by the hub and query a
+// catchup from the new server
+func (h *Hub) AddServerSocket(socket socket.Socket) error {
```
Comment (nkcr) [inline be1-go/hub/standard_hub/mod.go:186]: I would rename it to something like `NotifyNewServer`. There we can think that it only takes note of a new server, where it in fact also catchup from it.

File: be1-go/hub/standard_hub/message_handling.go [hunk 20]
```diff
@@ -0,0 +1,286 @@
...
+		publish := method.Publish{
+			Base: query.Base{
+				JSONRPCBase: jsonrpc.JSONRPCBase{
+					JSONRPC: "2.0",
+				},
+				Method: "publish",
+			},
+
+			Params: struct {
+				Channel string          "json:\"channel\""
```
Comment (nkcr) [inline be1-go/hub/standard_hub/message_handling.go:137]:
```suggestion
				Channel string          `json:"channel"`
```

Comment (nkcr) [reply be1-go/hub/standard_hub/message_handling.go:137]: same bellow

File: be1-go/hub/standard_hub/mod.go [hunk 21]
```diff
@@ -291,120 +317,82 @@ func (h *Hub) handleMessageFromClient(incomingMessage *socket.IncomingMessage) e
...
+	// check type (answer or query)
+	if rpctype == jsonrpc.RPCTypeAnswer {
+		err = h.handleAnswer(socket, byteMessage)
+		if err != nil {
+			err = answer.NewErrorf(-4, "failed to handle answer message: %v", err)
+			socket.SendError(nil, err)
+			return err
+		}
+		return nil
+	} else if rpctype != jsonrpc.RPCTypeQuery {
```
Comment (nkcr) [inline be1-go/hub/standard_hub/mod.go:349]: Useless `else if`:

```go
		}

		return nil
	}

	if rpctype != jsonrpc.RPCTypeQuery {
```

File: be1-go/hub/standard_hub/mod.go [hunk 22]
```diff
@@ -418,29 +406,42 @@ func (h *Hub) handleIncomingMessage(incomingMessage *socket.IncomingMessage) err
...
+func (h *Hub) broadcastToServers(message message.Message, byteMessage []byte) bool {
+	h.Lock()
+	defer h.Unlock()
+	_, ok := h.hubInbox.GetMessage(message.MessageID)
+	if !ok {
+		h.serverSockets.SendToAll(byteMessage)
+		h.hubInbox.StoreMessage(message)
+		return false
+	}
+	return true
```
Comment (nkcr) [inline be1-go/hub/standard_hub/mod.go:428]:
```suggestion
	h.Lock()
	defer h.Unlock()

	_, ok := h.hubInbox.GetMessage(message.MessageID)
	if !ok {
		h.serverSockets.SendToAll(byteMessage)
		h.hubInbox.StoreMessage(message)
		return false
	}
	
	return true
```

**Review by nkcr — CHANGES_REQUESTED** [review]
This PR is quite big with a lot of changes. I did my best but I'm might have missed some elements. 
Next time please try to come with something smaller and more scoped. For example reorganizing files and making changes in them at the same time makes the review difficult.

File: be1-go/message/query/method/answer.go [hunk 23]
```diff
@@ -0,0 +1,17 @@
+package method
```
[the remark this answers is not in the document]
Comment (Anghul) [reply be1-go/message/query/method/answer.go:1]: All files in the method folder are in the package method. It would make sense to change it, but I think it is better to change it in another PR.

[the remark this answers is not in the document]
Comment (nkcr) [reply be1-go/message/query/method/answer.go:1]: Sorry I still don't get why you are redefining an `Answer` type, while there is an existing one in `message/answer.go`.

File: be1-go/hub/mod.go [hunk 24]
```diff
@@ -23,6 +23,9 @@ const (
 // Hub defines the methods a PoP server must implement to receive messages
 // and handle clients.
 type Hub interface {
+	// AddServerSocket add a socket for the hub to send message to other servers
```
Comment (nkcr) [inline be1-go/hub/mod.go:26]:
```suggestion
	// NotifyNewServer add a socket for the hub to send message to other servers
```

**Review by nkcr — CHANGES_REQUESTED** [review]
A last change and it's good to go!

**Review by nkcr — APPROVED** [review]

File: be1-go/message/query/method/answer.go (hunk 23 above)
[the remark this answers is not in the document]
Comment (Anghul) [reply be1-go/message/query/method/answer.go:1]: I didn't understand what you were asking at first. So yeah, there is no specific reason, I guess I just missde the class. And it could (and should) be done (I already looked at it a bit), I just have to find a way to rewrite the tests as they are posing me a few problems because of the json schema validation. Would it be better to do it now, or to write an issue that I will work on later ?

[the remark this answers is not in the document]
Comment (nkcr) [reply be1-go/message/query/method/answer.go:1]: If its not a huge work then please include that in this PR.

**Review by nkcr — APPROVED** [review]
